File: src/rff.jl
# ==============================================================================
# Random Fourier Features (RFF) for MolInvDistSE
# ==============================================================================
#
# MolInvDistSE is an SE kernel in inverse-distance feature space:
#   k(x,y) = sigma^2 * exp(-theta^2 * ||phi(x) - phi(y)||^2)
#
# By Bochner's theorem, this has a Gaussian spectral density:
#   p(w) = N(0, 2*theta^2 * I)
#
# RFF approximates the kernel with D_rff random features:
#   k(x,y) ~ z(x)^T z(y)
# where z(x) = sigma * sqrt(2/D_rff) * cos(W * phi(x) + b).
#
# Training is Bayesian linear regression: O(N * D * D_rff + D_rff^3),
# which replaces the exact GP cost of O((N*(D+1))^3). For N=400 and
# D_rff=200 this is a ~1000x speedup.
#
# Gradient observations are incorporated via the Jacobian of z w.r.t. x,
# computed analytically using the chain rule through the inverse distance
# feature map.
#
# References:
#   Rahimi & Recht (2007), "Random Features for Large-Scale Kernel Machines"
#   Lazaro-Gredilla et al. (2010), "Sparse Spectrum GP Regression"

"""
    RFFModel

Random Fourier Features approximation to the MolInvDistSE GP.

Hyperparameters (sigma, theta) come from a base kernel trained on
a subset. Training and prediction use all N data points via the
D_rff-dimensional feature approximation.

# Fields
- `W`: (D_rff, d_feat) frequency matrix sampled from N(0, 2*theta^2*I)
- `b`: (D_rff,) random phase offsets in [0, 2pi)
- `c`: sigma * sqrt(2/D_rff) scaling constant
- `frozen`: flat coordinates of frozen atoms
- `alpha`: (D_rff,) Bayesian linear regression weights
- `A_chol`: Cholesky of the regularized Gram matrix (for variance)
"""
struct RFFModel
    W::Matrix{Float64}
    b::Vector{Float64}
    c::Float64
    frozen::Vector{Float64}
    alpha::Vector{Float64}
    A_chol::LinearAlgebra.Cholesky{Float64, Matrix{Float64}}
end

"""
    _rff_features(W, b, c, frozen, x) -> (z, J_z)

Compute RFF feature vector z(x) and its Jacobian J_z = dz/dx.

Returns:
- `z`: (D_rff,) feature vector
- `J_z`: (D_rff, D) Jacobian matrix
"""
function _rff_features(
    W::Matrix{Float64},
    b::Vector{Float64},
    c::Float64,
    frozen::Vector{Float64},
    x::AbstractVector{Float64},
)
    phi = compute_inverse_distances(x, frozen)
    u = W * phi .+ b
    z = c .* cos.(u)

    # Jacobian of inverse distances w.r.t. Cartesian coordinates
    J_phi = ForwardDiff.jacobian(
        x_ -> compute_inverse_distances(x_, frozen), collect(x))
    # J_phi: (d_feat, D), W: (D_rff, d_feat)
    # J_z = -c * sin(u) .* (W * J_phi), element-wise broadcast on columns
    sin_u = sin.(u)
    J_z = (-c) .* sin_u .* (W * J_phi)  # (D_rff, D)

    return z, J_z
end

"""
    build_rff(kernel, X_train, y_train, D_rff; noise_var, grad_noise_var) -> RFFModel

Build an RFF model from a trained MolInvDistSE kernel and all training data.

The frequency matrix W is sampled from the spectral density of the
trained kernel. Training solves a Bayesian linear regression problem
in the D_rff-dimensional feature space, cost O(N*D*D_rff + D_rff^3).

`y_train` must be in blocked layout: [E1..EN, G1_1..GN_D], matching
the output of `build_full_covariance`.
"""
function build_rff(
    kernel::MolInvDistSE,
    X_train::Matrix{Float64},
    y_train::Vector{Float64},
    D_rff::Int;
    noise_var::Float64 = 1e-6,
    grad_noise_var::Float64 = 1e-4,
)
    D, N = size(X_train)
    frozen = Vector{Float64}(kernel.frozen_coords)
    n_atoms = div(D, 3)
    n_frozen = div(length(frozen), 3)
    d_feat = div(n_atoms * (n_atoms - 1), 2) + n_atoms * n_frozen

    # Sample frequencies from spectral density N(0, 2*theta^2 * I)
    theta = kernel.inv_lengthscales
    W = zeros(D_rff, d_feat)
    if isempty(kernel.feature_params_map) && length(theta) == 1
        W .= randn(D_rff, d_feat) .* (sqrt(2.0) * theta[1])
    else
        for f in 1:d_feat
            idx = isempty(kernel.feature_params_map) ? 1 : kernel.feature_params_map[f]
            W[:, f] .= randn(D_rff) .* (sqrt(2.0) * theta[idx])
        end
    end

    b = rand(D_rff) .* (2.0 * pi)
    sigma = sqrt(kernel.signal_variance)
    c = sigma * sqrt(2.0 / D_rff)

    # Build design matrix Z in blocked layout: [energy rows; gradient rows]
    n_obs = N * (1 + D)
    Z = zeros(n_obs, D_rff)

    for i in 1:N
        xi = view(X_train, :, i)
        z, J_z = _rff_features(W, b, c, frozen, xi)

        # Energy row i
        Z[i, :] = z

        # Gradient rows: N + (i-1)*D + d for each Cartesian component d
        for d in 1:D
            Z[N + (i - 1) * D + d, :] = J_z[:, d]
        end
    end

    # Noise precision vector
    prec = zeros(n_obs)
    prec[1:N] .= 1.0 / noise_var
    prec[N+1:end] .= 1.0 / grad_noise_var

    # Bayesian linear regression: A = Z^T diag(prec) Z + I
    ZtP = Z' .* prec'   # (D_rff, n_obs)
    A = Symmetric(ZtP * Z + I(D_rff))
    A_chol = cholesky(A)

    alpha = A_chol \ (ZtP * y_train)

    return RFFModel(W, b, c, frozen, alpha, A_chol)
end

"""
    predict(rff::RFFModel, X_test) -> Vector

RFF prediction with interleaved output layout:
[E1, G1_1, ..., G1_D, E2, G2_1, ..., G2_D, ...].

Cost: O(D_rff * D) per test point.
"""
function predict(rff::RFFModel, X_test::Matrix{Float64})
    D, N_test = size(X_test)
    dim_block = D + 1
    result = zeros(N_test * dim_block)

    for i in 1:N_test
        xi = view(X_test, :, i)
        z, J_z = _rff_features(rff.W, rff.b, rff.c, rff.frozen, xi)

        offset = (i - 1) * dim_block
        result[offset + 1] = dot(z, rff.alpha)
        result[offset+2 : offset+dim_block] = J_z' * rff.alpha
    end

    return result
end

"""
    predict_with_variance(rff::RFFModel, X_test) -> (mean, variance)

RFF prediction with Bayesian linear regression variance.

Energy variance: var(E) = z^T A^{-1} z
Gradient variance: var(G_d) = J_z[:, d]^T A^{-1} J_z[:, d]

Output layout matches `predict_with_variance(::GPModel)`.
"""
function predict_with_variance(rff::RFFModel, X_test::Matrix{Float64})
    D, N_test = size(X_test)
    dim_block = D + 1
    mu = zeros(N_test * dim_block)
    var = zeros(N_test * dim_block)

    for i in 1:N_test
        xi = view(X_test, :, i)
        z, J_z = _rff_features(rff.W, rff.b, rff.c, rff.frozen, xi)

        offset = (i - 1) * dim_block

        # Energy mean + variance
        mu[offset + 1] = dot(z, rff.alpha)
        v = rff.A_chol.L \ z
        var[offset + 1] = dot(v, v)

        # Gradient mean + variance
        G = J_z' * rff.alpha
        for d in 1:D
            mu[offset + 1 + d] = G[d]
            v_d = rff.A_chol.L \ view(J_z, :, d)
            var[offset + 1 + d] = dot(v_d, v_d)
        end
    end

    return mu, var
end

File: crates/chemgp-core/src/rff.rs
//! Random Fourier Features for GP kernels.
//!
//! Ports `rff.jl`. Approximates the GP with O(N*D*D_rff + D_rff^3) cost.

use crate::kernel::Kernel;
use faer::linalg::solvers::Solve;
use faer::{Mat, Side};
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use rand_distr::StandardNormal;

/// How RFF features are computed from raw coordinates.
#[derive(Debug, Clone)]
pub enum FeatureMode {
    /// Inverse interatomic distances (molecular kernel).
    InverseDistances { frozen: Vec<f64> },
    /// Raw coordinates (Cartesian kernel, identity Jacobian).
    Cartesian,
}

/// RFF model for fast GP prediction.
pub struct RffModel {
    pub w: Mat<f64>,      // (d_rff, d_feat)
    pub b: Vec<f64>,      // (d_rff,)
    pub c: f64,           // sigma * sqrt(2/d_rff)
    pub feature_mode: FeatureMode,
    pub alpha: Vec<f64>,  // (d_eff,) where d_eff = d_rff + 1 (constant feature)
    pub a_chol: Mat<f64>, // Lower-triangular Cholesky factor of regularized Gram
    pub dim: usize,       // Coordinate dimension D
    pub const_sigma2: f64, // Constant kernel variance
}

impl RffModel {
    /// Compute RFF features z(x) and Jacobian J_z = dz/dx.
    fn features(&self, x: &[f64]) -> (Vec<f64>, Mat<f64>) {
        let d = x.len();
        let d_rff = self.w.nrows();

        let (phi, j_phi) = match &self.feature_mode {
            FeatureMode::InverseDistances { frozen } => {
                crate::kernel::invdist_jacobian(x, frozen)
            }
            FeatureMode::Cartesian => {
                // Features = coordinates, Jacobian = identity
                let phi = x.to_vec();
                let d_feat = d;
                let mut j_phi = Mat::<f64>::zeros(d_feat, d);
                for i in 0..d_feat {
                    j_phi[(i, i)] = 1.0;
                }
                (phi, j_phi)
            }
        };

        let d_feat = phi.len();

        // u = W * phi + b
        let mut u = vec![0.0; d_rff];
        for i in 0..d_rff {
            let mut s = self.b[i];
            for f in 0..d_feat {
                s += self.w[(i, f)] * phi[f];
            }
            u[i] = s;
        }

        let mut z: Vec<f64> = u.iter().map(|&v| self.c * v.cos()).collect();

        // w_jphi = W * J_phi (d_rff x d)
        let mut w_jphi = Mat::<f64>::zeros(d_rff, d);
        for i in 0..d_rff {
            for j in 0..d {
                let mut s = 0.0;
                for f in 0..d_feat {
                    s += self.w[(i, f)] * j_phi[(f, j)];
                }
                w_jphi[(i, j)] = s;
            }
        }

        // J_z = -c * sin(u) .* (W * J_phi)
        // Append constant feature: z gets sqrt(const_sigma2), j_z gets zero row.
        let d_eff = d_rff + 1;
        let mut j_z = Mat::<f64>::zeros(d_eff, d);
        for i in 0..d_rff {
            let sin_u = u[i].sin();
            for j in 0..d {
                j_z[(i, j)] = -self.c * sin_u * w_jphi[(i, j)];
            }
        }
        // Row d_rff (constant feature) is already zeros.

        z.push(self.const_sigma2.sqrt());

        (z, j_z)
    }
}

/// Build an RFF model from a trained kernel and all training data.
pub fn build_rff(
    kernel: &Kernel,
    x_train: &[f64],
    dim: usize,
    n: usize,
    y_train: &[f64],
    d_rff: usize,
    noise_var: f64,
    grad_noise_var: f64,
    seed: u64,
    const_sigma2: f64,
) -> RffModel {
    let inv_ls = kernel.inv_lengthscales();
    let d_feat = kernel.n_features(dim);

    let feature_mode = match kernel {
        Kernel::MolInvDist(k) => FeatureMode::InverseDistances {
            frozen: k.frozen_coords.clone(),
        },
        Kernel::Cartesian(_) => FeatureMode::Cartesian,
    };

    let mut rng = StdRng::seed_from_u64(seed);

    // Sample frequencies from N(0, 2*theta^2 * I)
    let feat_map = kernel.feature_params_map();
    let mut w = Mat::<f64>::zeros(d_rff, d_feat);
    if feat_map.is_empty() && inv_ls.len() == 1 {
        let scale = (2.0f64).sqrt() * inv_ls[0];
        for i in 0..d_rff {
            for f in 0..d_feat {
                w[(i, f)] = rng.sample::<f64, _>(StandardNormal) * scale;
            }
        }
    } else {
        for f in 0..d_feat {
            let idx = if feat_map.is_empty() { 0 } else { feat_map[f] };
            let scale = (2.0f64).sqrt() * inv_ls[idx];
            for i in 0..d_rff {
                w[(i, f)] = rng.sample::<f64, _>(StandardNormal) * scale;
            }
        }
    }

    let mut b = vec![0.0; d_rff];
    for i in 0..d_rff {
        b[i] = rng.random::<f64>() * 2.0 * std::f64::consts::PI;
    }

    let sigma = kernel.signal_variance().sqrt();
    let c = sigma * (2.0 / d_rff as f64).sqrt();

    // Build design matrix Z: [energy rows; gradient rows]
    // d_eff = d_rff + 1: extra column for constant kernel feature.
    let d_eff = d_rff + 1;
    let n_obs = n * (1 + dim);
    let mut z_mat = Mat::<f64>::zeros(n_obs, d_eff);

    let model = RffModel {
        w: w.clone(),
        b: b.clone(),
        c,
        feature_mode: feature_mode.clone(),
        alpha: vec![0.0; d_eff],
        a_chol: Mat::<f64>::zeros(d_eff, d_eff),
        dim,
        const_sigma2,
    };

    for i in 0..n {
        let xi = &x_train[i * dim..(i + 1) * dim];
        let (z, j_z) = model.features(xi);

        // Energy row i (d_eff features including constant)
        for f in 0..d_eff {
            z_mat[(i, f)] = z[f];
        }

        // Gradient rows (constant feature derivative is zero)
        for d in 0..dim {
            for f in 0..d_eff {
                z_mat[(n + i * dim + d, f)] = j_z[(f, d)];
            }
        }
    }

    // Noise precision
    let mut prec = vec![0.0; n_obs];
    for i in 0..n {
        prec[i] = 1.0 / noise_var;
    }
    for i in n..n_obs {
        prec[i] = 1.0 / grad_noise_var;
    }

    // A = Z^T diag(prec) Z + I
    // ztp = Z^T * diag(prec)  (d_eff x n_obs)
    let mut ztp = Mat::<f64>::zeros(d_eff, n_obs);
    for i in 0..d_eff {
        for j in 0..n_obs {
            ztp[(i, j)] = z_mat[(j, i)] * prec[j];
        }
    }

    // a = ztp * z_mat + I  (d_eff x d_eff)
    let mut a = Mat::<f64>::zeros(d_eff, d_eff);
    for i in 0..d_eff {
        for j in 0..d_eff {
            let mut s = 0.0;
            for k in 0..n_obs {
                s += ztp[(i, k)] * z_mat[(k, j)];
            }
            a[(i, j)] = s;
        }
        a[(i, i)] += 1.0;
    }

    let llt = a.llt(Side::Lower).expect("RFF Cholesky failed");
    let a_chol_l = {
        let l = llt.L();
        let mut m = Mat::<f64>::zeros(d_eff, d_eff);
        for i in 0..d_eff {
            for j in 0..=i {
                m[(i, j)] = l[(i, j)];
            }
        }
        m
    };

    // rhs = ztp * y  (d_eff x 1)
    let mut rhs = Mat::<f64>::zeros(d_eff, 1);
    for i in 0..d_eff {
        let mut s = 0.0;
        for j in 0..n_obs {
            s += ztp[(i, j)] * y_train[j];
        }
        rhs[(i, 0)] = s;
    }

    let alpha_mat = llt.solve(&rhs);
    let alpha: Vec<f64> = (0..d_eff).map(|i| alpha_mat[(i, 0)]).collect();

    RffModel {
        w,
        b,
        c,
        feature_mode,
        alpha,
        a_chol: a_chol_l,
        dim,
        const_sigma2,
    }
}

/// RFF prediction with interleaved output: [E1, G1_1..G1_D, E2, ...].
pub fn rff_predict(rff: &RffModel, x_test: &[f64], n_test: usize) -> Vec<f64> {
    let d = rff.dim;
    let dim_block = d + 1;
    let mut result = vec![0.0; n_test * dim_block];

    for i in 0..n_test {
        let xi = &x_test[i * d..(i + 1) * d];
        let (z, j_z) = rff.features(xi);

        let offset = i * dim_block;
        // Energy: z . alpha
        result[offset] = z.iter().zip(rff.alpha.iter()).map(|(a, b)| a * b).sum();
        // Gradient: J_z^T * alpha
        for dd in 0..d {
            let mut s = 0.0;
            for f in 0..z.len() {
                s += j_z[(f, dd)] * rff.alpha[f];
            }
            result[offset + 1 + dd] = s;
        }
    }

    result
}

/// RFF prediction with variance.
pub fn rff_predict_with_variance(
    rff: &RffModel,
    x_test: &[f64],
    n_test: usize,
) -> (Vec<f64>, Vec<f64>) {
    let d = rff.dim;
    let d_rff = rff.alpha.len();
    let dim_block = d + 1;
    let mut mu = vec![0.0; n_test * dim_block];
    let mut var = vec![0.0; n_test * dim_block];

    for i in 0..n_test {
        let xi = &x_test[i * d..(i + 1) * d];
        let (z, j_z) = rff.features(xi);

        let offset = i * dim_block;

        // Energy mean
        mu[offset] = z.iter().zip(rff.alpha.iter()).map(|(a, b)| a * b).sum();

        // Energy variance: v = L^{-1} z, var = v^T v
        let mut v = Mat::from_fn(d_rff, 1, |r, _| z[r]);
        rff.a_chol.solve_lower_triangular_in_place(&mut v);
        let mut vdot = 0.0;
        for r in 0..d_rff {
            vdot += v[(r, 0)] * v[(r, 0)];
        }
        var[offset] = vdot;

        // Gradient mean and variance
        for dd in 0..d {
            let mut s = 0.0;
            for f in 0..d_rff {
                s += j_z[(f, dd)] * rff.alpha[f];
            }
            mu[offset + 1 + dd] = s;

            let mut v_d = Mat::from_fn(d_rff, 1, |r, _| j_z[(r, dd)]);
            rff.a_chol.solve_lower_triangular_in_place(&mut v_d);
            let mut vd_dot = 0.0;
            for r in 0..d_rff {
                vd_dot += v_d[(r, 0)] * v_d[(r, 0)];
            }
            var[offset + 1 + dd] = vd_dot;
        }
    }

    (mu, var)
}
